### DISCUSSION TITLE
Specmatic with wiremock stubs

### DISCUSSION THREAD

**Discussion by gurubabujampala** [specmatic/specmatic#1362] [General]
Hi Team,

I had some time to watch the video [WireMock's Dirty Secret](https://www.youtube.com/watch?v=6V32q1I2VCs) From what I understand, WireMock cannot assist with the schema validation specified in the OpenAPI documentation.

I spent some time looking into this and found a few external libraries that can bridge the gap in WireMock

Swagger Request Validator Library : [swagger-request-validator-wiremock-junit5](https://bitbucket.org/atlassian/swagger-request-validator/src/master/swagger-request-validator-wiremock-junit5/)

Below is the Specmatic example with wiremock I have tried

```java
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class ContractWithWiremockTest extends SpecmaticJUnitSupport {

    @LocalServerPort
    private int port;

    private static final OpenApiValidator OPEN_API_VALIDATOR = new OpenApiValidator("visitinfo.yaml");
    private static final WireMockServer mockServer = new WireMockServer(options().port(9090)
            .extensions(new ResponseTemplateTransformer(true), OPEN_API_VALIDATOR));

    @BeforeAll
    static void setup() {
        System.setProperty("SPECMATIC_GENERATIVE_TESTS", "false");
       mockServer.start();
    }

    @AfterEach
    void validateWiremockInteractions() {
        // Validate WireMock server interactions with the provider OpenAPI specification
        OPEN_API_VALIDATOR.assertValidationPassed();
    }

    @BeforeEach
    public void setPort() {
        System.setProperty("host", "localhost");
        System.setProperty("port", String.valueOf(port));
        System.setProperty("endpointsAPI", "http://localhost:" + port+"/actuator/mappings");
        OPEN_API_VALIDATOR.reset();
    }

    @AfterAll
    public static void tearDown() throws IOException {
        mockServer.stop();
    }

}

``` 

Wiremock mappings stub file:
```javascript
{
    "request": {
        "url": "/api/visits/1001",
        "method": "GET"
    },
    "response": {
        "status": 200,
        "body": "{\"id\": \"1\", \"petId\": 1001, \"visited\": true, \"lastVisitedDate\": \"2022-10-11\"}",
        "headers": {
            "Content-Type": "application/json",
            "Connection": "keep-alive"
        }
    }
}
``` 
Open API Specification: 
```javascript
openapi: 3.0.1
info:
  title: OpenAPI definition
  version: v0
servers:
  - url: http://localhost:9090
    description: Generated server url
paths:
  /api/visits/{petId}:
    get:
      tags:
        - visit-controller
      operationId: visitInfo
      parameters:
        - name: petId
          in: path
          required: true
          schema:
            type: integer
            format: int32
          examples:
            1001:
              summary: ID 1001 example
              value: 1001
            1002:
              summary: ID 1002 example
              value: 1002
      responses:
        "200":
          description: OK
          content:
            'application/json':
              schema:
                $ref: "#/components/schemas/VisitDTO"
              examples:
                1001:
                  summary: Example with id 1001
                  value:
                    id: 1
                    petId: 1001
                    visited: true
                    lastVisitedDate: "2022-10-11"
                    clinicName: "Ramo Pet-Clinic"
                1002:
                  summary: Example with id 1002
                  value:
                    id: 2
                    petId: 1002
                    visited: false
                    clinicName: "Ramo Pet-Clinic"
        "400": # Adding the 400 response
          description: Bad Request
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: number
                  error:
                    type: string
                  path:
                    type: string
                  timestamp:
                    type: string
              examples: # Example for 400 Bad Request
                invalidId:
                  summary: Invalid ID provided
                  value:
                    status: 400
                    error: "Bad Request"
                    path: "path"
                    timestamp: "2024-09-23T02:55:09.533+00:00"

components:
  schemas:
    VisitDTO:
      type: object
      required:
        - id
        - petId
        - visited
        - clinicName
      properties:
        id:
          type: integer
          format: int32
        petId:
          type: integer
          format: int32
        lastVisitedDate:
          type: string
        visited:
          type: boolean
        clinicName:
          type: string

``` 

If a request or response doesn’t match the schema provided in the **_OAS, the OPEN_API_VALIDATOR.assertValidationPassed())_** validator will generate an error indicating that the response/request doesn’t conform to the schema, along with the error details.

In the above examples there are couple of gaps, First one is **_Id_** is **_Integer_** in OAS where as it's **_string_** in wiremock stub and the other one is **_clinicName_** is required in OAS but it's missing in the response from wiremock stub. If I run test I am seeing the below errors

-  **"message" : "Object has missing required properties ([\"clinicName\"])",**
-  **"message" : "[Path '/id'] Instance type (string) does not match any allowed primitive type (allowed: [\"integer\"])",**

**Next Steps:** 
We are currently exploring if there is a library that can create WireMock stub mapping files from an OAS, similar to how Specmatic generates stubs.

Kindly help are there any gaps apart this with wiremock when you were trying from specmatic? Also please let us know are there any libraries you come across to generate wiremock stubs mappings files from open api specification and you see the gaps during your initial exploration to build specmatic stub server?

**Comment by harikrishnan83** [comment]
Hello @gurubabujampala,

Apologies for the delay in responding, and thank you so much for taking the time to watch the "[WireMock's Dirty Secret](https://www.youtube.com/watch?v=6V32q1I2VCs)" video. I also appreciate you sharing about the validator library you’ve tried.

Before I address your question about other gaps in WireMock, I’d like to take a step back and understand the context. Can you please clarify why you are trying to use WireMock instead of Specmatic to mock the dependency? Could you please share if there’s any specific advantage WireMock offers you here?

This will really help me provide a more informed response to your question.
